apiserver/common/networkingcommon/networkconfigapi.go

```diff
@@ -117,29 +129,229 @@ func (api *NetworkConfigAPI) getMachineForSettingNetworkConfig(machineTag string
+	// will have removed the provider ID - see the instance-poller API facade
+	// logic.
+	if removing == len(addrs) && dev.ProviderID() == "" {
+		o.removalCandidates = append(o.removalCandidates, dev)
+	}
+
+	return ops, nil
+}
+
+func (o *updateMachineLinkLayerOp) processExistingDeviceAddress(
+	dev LinkLayerDevice, addr LinkLayerAddress, incomingAddrs []state.LinkLayerDeviceAddress,
+) ([]txn.Op, error) {
+	addrValue := addr.Value()
+
+	// If one of the incoming addresses matches the existing one,
+	// update it.
+	for _, incomingAddr := range incomingAddrs {
+		if strings.HasPrefix(incomingAddr.CIDRAddress, addrValue) &&
+			!o.IsAddrProcessed(dev.MACAddress(), incomingAddr.CIDRAddress) {
+			o.MarkAddrProcessed(dev.MACAddress(), incomingAddr.CIDRAddress)
+
+			ops, err := addr.UpdateOps(incomingAddr)
+			return ops, errors.Trace(err)
+		}
+	}
+
+	// Otherwise if we are the authority, delete it.
+	if addr.Origin() == network.OriginMachine {
+		logger.Debugf("removing address %q from device %q", addrValue, addr.DeviceName())
+		return addr.RemoveOps(), nil
+	}
+
+	return nil, nil
+}
+
+// processExistingDeviceNewAddresses interrogates the list of incoming
+// addresses and adds any that were not processed as already existing.
+func (o *updateMachineLinkLayerOp) processExistingDeviceNewAddresses(
+	dev LinkLayerDevice, incomingAddrs []state.LinkLayerDeviceAddress,
+) ([]txn.Op, error) {
+	var ops []txn.Op
+	for _, addr := range incomingAddrs {
+		if !o.IsAddrProcessed(dev.MACAddress(), addr.CIDRAddress) {
+			addOps, err := dev.AddAddressOps(addr)
+			if err != nil {
+				return nil, errors.Trace(err)
+			}
+			ops = append(ops, addOps...)
+
+			o.MarkAddrProcessed(dev.MACAddress(), addr.CIDRAddress)
+		}
+	}
+	return ops, nil
+}
+
+// processNewDevices handles incoming devices that
+// did not match any we already have in state.
+func (o *updateMachineLinkLayerOp) processNewDevices() ([]txn.Op, error) {
+	var ops []txn.Op
+	for _, dev := range o.Incoming() {
+		if !o.IsDevProcessed(dev) {
+			logger.Debugf("adding new device %q (%s) with addresses %v",
+				dev.InterfaceName, dev.MACAddress, dev.Addresses)
+
+			addOps, err := o.machine.AddLinkLayerDeviceOps(
+				networkDeviceToStateArgs(dev), o.MatchingIncomingAddrs(dev.MACAddress)...)
+			if err != nil {
+				return nil, errors.Trace(err)
+			}
+			ops = append(ops, addOps...)
+
+			o.MarkDevProcessed(dev.MACAddress)
+		}
+	}
+	return ops, nil
+}
+
+// processRemovalCandidates returns transaction operations for
+// removing unobserved devices that it is safe to delete.
+// A device is considered safe to delete if it has no children,
+// or if all of its children are also candidates for deletion.
+// Any device considered here will already have ops generated
+// for removing its addresses.
+func (o *updateMachineLinkLayerOp) processRemovalCandidates() []txn.Op {
+	var ops []txn.Op
+	for _, dev := range o.removalCandidates {
+		if o.observedParentDevices.Contains(dev.Name()) {
```

Comment:
Is it possible to have a parent - child topology where **both** devices get removed as part of this update operation? If that's the case then this check would only remove the child and not the parent. Of course, this will be eventually consistent (the parent will be removed in the next machiner update cycle) but it would be great to address such scenarios in a single update.

You might be able to work around this if you made `observedParentDevices` into a `map[string]int` where keys are dev names and values are then number of child devices with a particular device as a parent. Then, this loop would do a pass for the children first (names not present in the map) and for each removal, decrement the counter. A second pass would get rid of the parents IFF their count is set to zero (and warn if that's not the case).

Reply:
The observed parents is only populated from *incoming* devices. I.e. if an existing parent and its children are not observed, there would be no addition of the parent's device name, so they all would be deleted without issue.

It works similarly if 2 bonded NICs were unbound. They would no longer have a parent to go into the guard set - the NICs would be updated to remove the parent, and the parent would get deleted.

Reply:
Ah... missed that bit. Thanks for the clarification.